### PR TITLE
Refactor codebase by splitting  crawling functions into separate files closes #108

### LINKED ISSUE #108: Refactor the "crawl.py" file to improve maintainability and organisation
## Description:

The `crawl.py` file in our project contains 31 functions and over 500 lines of code. While this file may have started out as a convenient place to put all of our crawling functions, it will become unwieldy and difficult to maintain as the project grow so it such be fixed as early as possible.

I propose refactoring `crawl.py` by splitting the functions into separate files based on their relation and functionality. Here are some benefits of doing so:

- Improved organization: Breaking up the code into smaller, more focused files will make it easier to find and modify specific functions. It will also make the codebase easier to navigate for new team members.

- Easier testing: Smaller files with focused functions make it easier to test individual pieces of functionality. This will improve our testing coverage and make it easier to catch bugs before they reach production.

- Reduced complexity: The size and complexity of `crawl.py` makes it difficult to reason about the code and understand how it all fits together. By breaking up the functions into separate files, we can reduce the cognitive load required to work with the codebase. Basically making the code more beginner-friendly 

I propose that each file should contain a subset of related functions that perform a specific task. This will allow us to more easily reason about each file's purpose and the functions contained within it.

## Related Issues
#2

### PR DESCRIPTION
**Pull request by Am0stafa**
## Description

This pull request implements the proposal outlined in issue #108 to refactor the `crawl.py` file by splitting the functions into separate files based on their relation and functionality.

Additionally, I have written a doc for each file that describes the functions it contains and their purpose. I have also added an overview doc for the module.

The refactored code is more modular and easier to maintain, with each file containing a subset of related functions that perform a specific task. This improves organization, and testing, and reduces complexity, making the codebase easier to navigate for new team members.

This proposal is essential to improving the maintainability of our codebase, and the changes made in this pull request will have a significant impact on the overall quality of our code. 

In addition to the refactoring, I have also updated the test files to import from the new module created, ensuring that our test coverage remains intact.

## Changes Made

- Refactored `crawl.py` by splitting the functions into separate files based on their relation and functionality
- Updated test files to import from the new module created

## How to Test

- Clone this branch and run the test files using `python -m unittest discover` command.

## Related Issues

- Resolves #108

### CODE REVIEW

**Comment by Am0stafa** [comment]
@mshudrak  I think the tests are failing because they need to be changed similar to what I did in the test files.
Or what do you think?

**Comment by Am0stafa** [comment]
I was initially going to make it into an OOP solution but decided to leave it to GSOC when I get accepted hopefully,
and by this change, it can be easily refactored.

**Comment by Am0stafa** [comment]
Also will help in adding more GCP resources project and after this change, any new resource added will follow best practices.

File: src/gcp_scanner/gcp_crawlers/__main__.py
Comment (ZetaTwo) [inline src/gcp_scanner/gcp_crawlers/__main__.py:1]: Why an empty file?

File: src/gcp_scanner/gcp_crawlers/__init__.py
Comment (ZetaTwo) [inline src/gcp_scanner/gcp_crawlers/__init__.py:15]: Add newline to end

**Review by ZetaTwo — CHANGES_REQUESTED** [review]
Overall looks good. Is it safe to assume that you did not change any code inside any of the functions and only rearranged the files?

File: src/gcp_scanner/gcp_crawlers/__main__.py
Comment (Am0stafa) [reply src/gcp_scanner/gcp_crawlers/__main__.py:1]: I created the main.py file as a potential entry point for the module, but currently, it is not being utilized. However, I believe that having it in place could be beneficial in the future. Additionally, there is a similar empty main.py file in the GCP scanner file, which I assume was created for the same purpose.

Comment (Am0stafa) [reply src/gcp_scanner/gcp_crawlers/__main__.py:1]: However, if you feel that it's not currently serving a purpose in the file, I could remove it for now and add it back in if and when it becomes necessary. The same goes for the empty file in the GCP scanner.

**Comment by Am0stafa** [comment]
> Overall looks good. Is it safe to assume that you did not change any code inside any of the functions and only rearranged the files?

Yes, that's correct. I did not change any code inside the functions themselves, only rearranged the files to make them more organized and easier to maintain.

File: src/gcp_scanner/gcp_crawlers/__main__.py
Comment (ZetaTwo) [inline src/gcp_scanner/gcp_crawlers/__main__.py:1]: Let's delete this for now

**Review by ZetaTwo — CHANGES_REQUESTED** [review]
Let's delete the empty file. Also, the tests are currently failing. Why is this?

File: src/gcp_scanner/gcp_crawlers/__main__.py
Comment (Am0stafa) [reply src/gcp_scanner/gcp_crawlers/__main__.py:1]: Understood

**Comment by Am0stafa** [comment]
> Let's delete the empty file. Also, the tests are currently failing. Why is this?

I believe that the failing tests may be due to the recent changes made to the project structure. One possible reason could be related to the import statements used in the test files. I recommend updating the import statements to match the new project structure, similar to what you did in the test file. This should help resolve the failing tests. What are your thoughts on this?

**Comment by Am0stafa** [comment]
Upon further investigation, I realized that removing the __init__.py file from the main directory caused errors and shouldn't delete, but now I understood why its there.

**Comment by ZetaTwo** [comment]
Ok, please make sure the tests are still working before we can merge this

**Comment by Am0stafa** [comment]
> Ok, please make sure the tests are still working before we can merge this

I fixed the testing errors in this PR by changing the import statement as see but idk why they arent passing when they are running using github actions

**Comment by Am0stafa** [comment]
@ZetaTwo working on fixing the last error

**Comment by Am0stafa** [comment]
@ZetaTwo, I have resolved all the failing tests except for the one related to the `list_services function`. Although the function remains the same, I am still investigating the cause of the failure. Do you have any insights or suggestions that could help me debug this issue?

**Comment by Am0stafa** [comment]
Also, @mshudrak any suggestion/guide as to why this test is failing?

**Comment by mshudrak** [comment]
The `services` unit_test is failing because expected output from GCP has changed. Please do not alter the test. All you need is to identify what changed and update tests/services file accordingly.

**Comment by Am0stafa** [comment]
> 

As I understand the list_services unit test only checks if there is output using the  and all I did was achieved the same goal but using a different method 

Thank you for the feedback, I appreciate it. I understand that the expected output from GCP has changed and that we should not alter the test case. My intention with the modification I made to the verify function was simply to ensure that the assertTrue assertion method passed when the list_services method returned a non-empty list which is exactly what the initial test did but in a different way, the initial test used the verify function is being used to check if the list of services returned by list_services contains at least one service

If updating the services file is the best course of action, I will certainly do so. However, if the only change needed is to modify the verify function to correctly check for the expected output, I believe this would be a simpler solution.

**Comment by Am0stafa** [comment]
@mshudrak Please let me know your thoughts on this, and if there is anything else I can do to help resolve the issue.

**Comment by mshudrak** [comment]
We want to check exact output from GCP with what we have in our tests. There is issue regarding relaxing the unit_tests but it is a separate problem that requires a lot of work. For now, we just want to check exact output from GCP in unit_tests.

**Comment by Am0stafa** [comment]
> There is issue regarding relaxing the unit_tests but it is a separate problem that requires a lot of work. For now, we just want to check exact output from GCP in unit_tests.

yeah I get that, I reverted back the unit test change and now it's falling but by going through the output of the `list_services` and the service JSON they are identical as I can see the only difference I notest is the single quote and double quote

Screen Shot 2023-03-17 at 4 27 36 PM

**Comment by Am0stafa** [comment]
@mshudrak I also remember that on one of my previous PRs the `list_services` was also falling I think in a similar way and I think you solved it

**Comment by Am0stafa** [comment]
@mshudrak thank you so much for the guidance btw

**Comment by mshudrak** [comment]
I think it complains about 
```
----------------------------- Captured stdout call -----------------------------
Retrieving access token from instance metadata
Successfully retrieved instance metadata
The following line was not identified in the output:
        "title": "sqladmin API (prod)",
```

**Comment by mshudrak** [comment]
Just ignore this test, I will fix it later.

**Comment by mshudrak** [comment]
So far we have 2 PRs for refactoring scanner. I think we need to have some generic class for objects creation as well as specific classes for specific GCP resources. Overall, I believe it is the move in the right direction. However, I think it would be  better to have the right solution in the first place without intermediate commits.

Anyway, thanks for PR, I'd link that one in your GSoC application proposal as an example of what you want to achieve with refactoring. I'd also prefer to have some design, discussion with mentors and then implementation, so I'd leave that one for GSoC timeframe.

**Comment by Am0stafa** [comment]
> So far we have 2 PRs for refactoring scanner. I think we need to have some generic class for objects creation as well as specific classes for specific GCP resources. Overall, I believe it is the move in the right direction. However, I think it would be better to have the right solution in the first place without intermediate commits.
> 
> Anyway, thanks for PR, I'd link that one in your GSoC application proposal as an example of what you want to achieve with refactoring. I'd also prefer to have some design, discussion with mentors and then implementation, so I'd leave that one for GSoC timeframe.

Thank you for your feedback. I understand your concerns about having the right solution in place without intermediate commits and having a more detailed design and discussion with the mentors before implementing changes. However, I believe that this initial refactoring can be a great starting point for future improvements, and it will make the project structure better and more maintainable. By merging this PR, we can create a solid foundation for the upcoming work and make it easier for both me and other contributors to build upon it. Would you consider merging this PR as a way of improving the project's overall structure and making it easier for future development?

**Comment by grumpyp** [comment]
What about a class based approach?

**Comment by Am0stafa** [comment]
> What about a class based approach? some ideas here #113

Actually i had implemented that but left it for gsoc timeframe

**Comment by Am0stafa** [comment]
changes made are for my gsoc proposal

File: src/gcp_scanner/gcp_crawlers/__init__.py
Comment (grumpyp) [inline src/gcp_scanner/gcp_crawlers/__init__.py:1]: wildcard imports are not best practise in my opinion, what's the reason you wouldn't import the modules themselfs like `import compute_cralwer` `import db_crawler` and so on?

> [Although certain modules are designed to export only names that follow certain patterns when you use import *, it is still considered bad practice in production code.](https://docs.python.org/3/tutorial/modules.html#importing-from-a-package)

**Review by grumpyp — CHANGES_REQUESTED** [review]

File: src/gcp_scanner/gcp_crawlers/__init__.py
Comment (Am0stafa) [reply src/gcp_scanner/gcp_crawlers/__init__.py:1]: Thanks, I really appreciate that you pointed that out, but I am aware of that and I had already mentioned that in my proposal along with other principles and best practices. In summary, I appreciate your input, and I am committed to following the best practices and principles for coding. Using explicit imports is one of these best practices, and I will ensure to adhere to it throughout my work.

**Comment by mshudrak** [comment]
I am closing this one for now but we can always reopen it if we decided to move forward with this refactoring attempt.

**Pull request closed without merging**
